components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttConnectionFactory.java

```diff
@@ -0,0 +1,283 @@
+                    passedInParameter.getProperty(MqttConstants.MQTT_SESSION_CLEAN));
+
+            parameters.put(MqttConstants.MQTT_SSL_ENABLE,
+                    passedInParameter.getProperty(MqttConstants.MQTT_SSL_ENABLE));
+
+        } catch (Exception e) {
+            log.error("Error while reading properties for MQTT Connection Factory " + factoryName,
+                    e);
+        }
+
+    }
+
+    public String getName() {
+        return factoryName;
+    }
+
+    public MqttClient getMqttClient() {
+        return createMqttClient();
+    }
+
+    public MqttAsyncClient getMqttAsyncClient() {
+        return createMqttAsyncClient();
+    }
+
+    public String getTopic() {
+        return parameters.get(MqttConstants.MQTT_TOPIC_NAME);
+    }
+
+    public String getContent() {
+        return parameters.get(MqttConstants.CONTENT_TYPE);
+    }
+
+
+    /**
+     * @return a synchronous MQTT client
+     */
+    private MqttClient createMqttClient() {
+
+        String uniqueClientId = MqttClient.generateClientId();
+
+        String sslEnable = parameters.get(MqttConstants.MQTT_SSL_ENABLE);
+
+        // This sample stores in a temporary directory... where messages
+        // temporarily
+        // stored until the message has been delivered to the server.
+        String tmpDir = parameters.get(MqttConstants.MQTT_TEMP_STORE);
+        MqttDefaultFilePersistence dataStore = null;
+
+        int qos = Integer.parseInt(parameters.get(MqttConstants.MQTT_QOS.toString()));
+        if (qos == 2 || qos == 1) {
+            if (tmpDir != null) {
+                dataStore = new MqttDefaultFilePersistence(tmpDir);
+            } else {
+                tmpDir = System.getProperty("java.io.tmpdir");//TODO when qos not needed?
+                dataStore = new MqttDefaultFilePersistence(tmpDir);
+            }
+        } else {
+            dataStore = null;
+        }
+
+        String mqttEndpointURL =
+                "tcp://" + parameters.get(MqttConstants.MQTT_SERVER_HOST_NAME) +
+                        ":" + parameters.get(MqttConstants.MQTT_SERVER_PORT);
+        // If SSL is enabled in the config, Use SSL tranport
+        if (sslEnable != null && sslEnable.equalsIgnoreCase("true")) {
+            mqttEndpointURL =
+                    "ssl://" + parameters.get(MqttConstants.MQTT_SERVER_HOST_NAME) + ":" +
+                            parameters.get(MqttConstants.MQTT_SERVER_PORT);
+        }
+
+        MqttClient mqttClient = null;
+        try {
+            if (dataStore != null) {
+                mqttClient = new MqttClient(mqttEndpointURL, uniqueClientId, dataStore);
+            } else {
+                mqttClient = new MqttClient(mqttEndpointURL, uniqueClientId);
+            }
+            log.info("Successfully created to mqtt client");
+        } catch (MqttException e1) {
+            int retryC = 0;
+            while ((mqttClient == null) && ((retryCount == -1) || (retryC < retryCount))) {
```

Comment:
We discussed that we need to put a thread sleep for each retry. Why is it missing?